### PR TITLE
Add resourceApply for VolumeSnapshotClass

### PR DESCRIPTION
**Pull request by Fedosin**
This commit allows managing VolumeSnapshotClass objects using a dynamic client.

### CODE REVIEW

File: pkg/operator/resource/resourceapply/volumesnapshotclass.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	requiredParameters, _, err := unstructured.NestedMap(required.UnstructuredContent(), "parameters")
+	if err != nil {
+		return nil, false, err
+	}
+	existingParameters, _, err := unstructured.NestedMap(existing.UnstructuredContent(), "parameters")
+	if err != nil {
+		return nil, false, err
+	}
+
+	if err := mergo.Merge(&existingParameters, &requiredParameters); err != nil {
```
Comment (deads2k) [inline pkg/operator/resource/resourceapply/volumesnapshotclass.go:41]: I suspect you want to just overwrite on any change.  I don't think you want half of your parameters and half a user-corrupted parameter.

File: pkg/operator/resource/resourceapply/volumesnapshotclass.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+		return nil, false, err
+	}
+
+	if equality.Semantic.DeepEqual(existingParameters, requiredParameters) {
+		return existing, false, nil
+	}
+
+	existingCopy := existing.DeepCopy()
+	if err := unstructured.SetNestedMap(existingCopy.UnstructuredContent(), existingParameters, "parameters"); err != nil {
```
Comment (deads2k) [inline pkg/operator/resource/resourceapply/volumesnapshotclass.go:50]: This was the typo we found in the spec version of this.  You want the requiredParameters here, right?

File: pkg/operator/resource/resourceapply/volumesnapshotclass.go (hunk 1 above)
Comment (jsafrane) [reply pkg/operator/resource/resourceapply/volumesnapshotclass.go:41]: I agree, overwrite everything.

File: pkg/operator/resource/resourceapply/volumesnapshotclass.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+			return nil, true, createErr
+		}
+		recorder.Eventf("VolumeSnapshotClassCreated", "Created VolumeSnapshotClass.snapshot.storage.k8s.io/v1 because it was missing")
+		return newObj, true, nil
+	}
+	if err != nil {
+		return nil, false, err
+	}
+
+	existingCopy := existing.DeepCopy()
```
Comment (jsafrane) [inline pkg/operator/resource/resourceapply/volumesnapshotclass.go:68]: Why copy the object here? It gets copied again in `ensureGenericVolumeSnapshotClass`

File: pkg/operator/resource/resourceapply/volumesnapshotclass.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+	toUpdate, modified, err := ensureGenericVolumeSnapshotClass(required, existingCopy)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if !modified {
+		return nil, false, nil
```
Comment (jsafrane) [inline pkg/operator/resource/resourceapply/volumesnapshotclass.go:76]: it should return `existing` here

File: pkg/operator/resource/resourceapply/volumesnapshotclass.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	requiredParameters, _, err := unstructured.NestedMap(required.UnstructuredContent(), "parameters")
+	if err != nil {
+		return nil, false, err
+	}
+	existingParameters, _, err := unstructured.NestedMap(existing.UnstructuredContent(), "parameters")
+	if err != nil {
+		return nil, false, err
+	}
+
+	if equality.Semantic.DeepEqual(existingParameters, requiredParameters) {
```
Comment (jsafrane) [inline pkg/operator/resource/resourceapply/volumesnapshotclass.go:40]: To be 100% sure you should check all fields, like `DeletionPolicy` and `Driver`

**Pull request merged**